xtm/xtm_api.cc

```diff
@@ -0,0 +1,282 @@
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "xtm_api.h"
+#include "xtm_scsp_queue.h"
+#include "xtm_config.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <errno.h>
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+#include <sys/eventfd.h>
+#endif
+
+#define XTM_PIPE_SIZE 4096
+
+union xtm_msg {
+	/**
+	 * Anonymous structure for storing functions and their arguments,
+	 * when using the xtm API in dispatch/invoke pattern.
+	 */
+	struct {
+		void (*fun)(void *);
+		void *fun_arg;
+	};
+	/**
+	 * Pointer to save message, when using xtm API
+	 * in send/recv pattern.
+	 */
+	void *data;
+};
+
+struct xtm_queue {
+	/**
+	 * File descriptor that the consumer thread must poll,
+	 * to know when new messages are added to the queue.
+	 */
+	int readfd;
+	/**
+	 * The file descriptor to which the producer thread writes
+	 * to notify the consumer thread of new messages in the queue.
+	 */
+	int writefd;
+	/** Message queue, it's size must be power of two */
+	struct xtm_scsp_queue<union xtm_msg> queue;
+};
+
+/**
+ * Function gets num data pointers, previously saved in data field of
+ * union xtm_msg. Used in case send/recv xtm API pattern.
+ * @param[in] queue - xtm queue
+ * @param[out] data - array to save data pointers from xtm_msg union data field
+ * @param[in] num - count of elements to get
+ * @retval return the number of data pointers actually read.
+ */
+static inline unsigned
+xtm_queue_get_data(struct xtm_queue *xtm_queue,
+		   void **data, unsigned num)
+{
+	struct xtm_scsp_queue<union xtm_msg> *queue = &xtm_queue->queue;
+	unsigned i = 0;
+	unsigned new_read = queue->read;
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+
+	for (i = 0; i < num; i++) {
+		if (new_read == queue_write)
+			break;
+		data[i] = queue->buffer[new_read].data;
+		new_read = (new_read + 1) & (queue->len - 1);
+	}
+	__atomic_store_n(&queue->read, new_read, __ATOMIC_RELEASE);
```

Comment:
Why do you touch internals of the queue manually? That should be a function in the queue API. Otherwise the encapsulation is broken.

Reply:
Fixed